### PR TITLE
Add ffs intrinsic

### PR DESCRIPTION
**Pull request by sbastrakov**
This is extension of #1004 , adding `ffs()` integer intrinsic.

### CODE REVIEW

**Comment by BenjaminW3** [comment]
May we want to add a MSVC version?
https://gitlab.freedesktop.org/cairo/cairo/commit/f5167dc2e1a13d8c4e5d66d7178a24b9b5e7ac7a

**Review by BenjaminW3 — DISMISSED** [review]

File: include/alpaka/intrinsic/Traits.hpp [hunk 1]
```diff
@@ -77,5 +84,51 @@ namespace alpaka
...
+            typename TIntrinsic>
+        ALPAKA_FN_ACC auto ffs(
+            TIntrinsic const & intrinsic,
+            long long value)
+        -> int
+        {
+            using ImplementationBase = concepts::ImplementationBase<ConceptIntrinsic, TIntrinsic>;
+            return traits::Ffs<
+                ImplementationBase>
+            ::ffs(
```
Comment (psychocoderHPC) [inline include/alpaka/intrinsic/Traits.hpp:130]: should this not be `ffsll()` and require the incude `<ctsring>` in this file see https://linux.die.net/man/3/ffsll

**Comment by psychocoderHPC** [comment]
general question: Under windows `int64_t` is `long int` under linux `long long int` but we have only a specialization for `int` and `long long int`. Do we ne need to add also `long int`? IMO `int64_t` should lead into an compile error (not tried I have no windows^^).

File: test/unit/intrinsic/src/Ffs.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    alpaka::test::KernelExecutionFixture<Acc> fixture(
+        alpaka::vec::Vec<Dim, Idx>::ones());
+
+    FfsTestKernel<int> kernel32bit;
+    REQUIRE(
+        fixture(
+            kernel32bit));
+
+    FfsTestKernel<long long> kernel64bit;
```
Comment (psychocoderHPC) [inline test/unit/intrinsic/src/Ffs.cpp]: IMO we should check `int64_t` and `int32_t`, Those types are typically used by the user and maps to `long long` or `long` but this is OS dependent.

**Review by psychocoderHPC — DISMISSED** [review]

**Comment by sbastrakov** [comment]
~~@BenjaminW3 MS version is signed, not unsigned, unlike GCC and CUDA/HIP. After spending some time on it, I do not see a super easy way to express it via pure intrinsic calls.~~

Edit: my comment above is wrong, it was the other way around with signed/unsigned and I see how they are expressing in the code you linked. I will add it.

File: test/unit/intrinsic/src/Ffs.cpp (hunk 2 above)
Comment (sbastrakov) [reply test/unit/intrinsic/src/Ffs.cpp]: That is a good point. The problem is that how GCC and CUDA define it is for these types, of not fixed width. I can make alpaka interface more modern C++ and `static_assert` internally that the type matches. Are you fine with it?

File: include/alpaka/intrinsic/Traits.hpp (hunk 1 above)
Comment (sbastrakov) [reply include/alpaka/intrinsic/Traits.hpp:130]: I followed the same logic as popcount: alpaka provides a unified interface (be it with templates or overloading like we discussed during review of #1004), without this ugly C postfixes
Edit: and this is call to the method, not that Linux function (which I do not use at all).

File: test/unit/intrinsic/src/Ffs.cpp (hunk 2 above)
Comment (sbastrakov) [reply test/unit/intrinsic/src/Ffs.cpp]: @psychocoderHPC I guess my last answer also relates to your [earlier comment](https://github.com/alpaka-group/alpaka/pull/1018#issuecomment-635957171), not just the one above.

**Comment by sbastrakov** [comment]
@BenjaminW3 @psychocoderHPC thanks for your comments. I added intrinsic-based implementation for Visual Studio, and tried to be more modern C++ with 32-bit vs. 64-bit types, and rebased.

File: include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp [hunk 3]
```diff
@@ -79,6 +79,30 @@ namespace alpaka
...
+            //#############################################################################
+            template<>
+            struct Ffs<
+                IntrinsicUniformCudaHipBuiltIn>
+            {
+                //-----------------------------------------------------------------------------
+                __device__ static auto ffs(
+                    intrinsic::IntrinsicUniformCudaHipBuiltIn const & /*intrinsic*/,
+                    int value)
+                -> int
```
Comment (psychocoderHPC) [inline include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp]: why do you not return a `std::int32_t` here too? To have the same return type you have on the CPU?

**Comment by psychocoderHPC** [comment]
@sbastrakov Please have a look to this CUDA CI issue: https://github.com/alpaka-group/alpaka/pull/1018/checks?check_run_id=730928384#step:4:6184

```
/home/runner/work/alpaka/alpaka/include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp:89:40: note: candidate function
                __device__ static auto ffs(
                                       ^
/home/runner/work/alpaka/alpaka/include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp:98:40: note: candidate function
                __device__ static auto ffs(
                                       ^
```

File: include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp (hunk 3 above)
Comment (sbastrakov) [reply include/alpaka/intrinsic/IntrinsicUniformCudaHipBuiltIn.hpp]: Yes, I should, thanks for pointing out.

**Review by BenjaminW3 — APPROVED** [review]

**Comment by sbastrakov** [comment]
Yes, comments of @psychocoderHPC were addressed. He is on vacation however.

**Comment by BenjaminW3** [comment]
@sbastrakov If your are done you can merge it.

**Pull request merged**